controllers/fenceagentsremediation_controller.go

```diff
@@ -337,52 +382,114 @@ func getNodeName(far *v1alpha1.FenceAgentsRemediation) string {
+}
+
+// buildFenceAgentParams collects the FAR's parameters for the node based on FAR CR, and if the CR is missing parameters
+// or the CR's name don't match nodeParameter name, or it has an action which is different from reboot, then return an error
+func buildFenceAgentParams(far *v1alpha1.FenceAgentsRemediation, secretParams map[string]string) ([]string, error) {
+	logger := ctrl.Log.WithName("build-fa-parameters")
 	var fenceAgentParams []string
-	// add shared parameters except the action parameter
+	//this map is used to verify there are not parameters which are defined twice
+	fenceAgentParamNames := make(map[v1alpha1.ParameterName]bool)
+
+	// append shared parameters
 	for paramName, paramVal := range far.Spec.SharedParameters {
-		if paramName != parameterActionName {
-			fenceAgentParams = appendParamToSlice(fenceAgentParams, paramName, paramVal)
-		} else if paramVal != parameterActionValue {
-			// --action attribute was selected but it is different than reboot
-			err := errors.New("FAR doesn't support any other action than reboot")
-			logger.Error(err, "can't build CR with this action attribute", "action", paramVal)
+		//Verify action must be reboot
+		if err := validateRebootAction(paramName, paramVal, logger); err != nil {
 			return nil, err
 		}
+		//Verify param isn't already defined
+		if err := validateUniqueParam(fenceAgentParamNames, paramName, logger); err != nil {
+			return nil, err
+		}
+		fenceAgentParams = appendParamToSlice(fenceAgentParams, paramName, paramVal)
 	}
-	// if --action attribute was not selected, then its default value is reboot
-	// https://github.com/ClusterLabs/fence-agents/blob/main/lib/fencing.py.py#L103
-	// Therefore we can safely add the reboot action regardless if it was initially added into the CR
-	fenceAgentParams = appendParamToSlice(fenceAgentParams, parameterActionName, parameterActionValue)
 
+	// append secret parameters
+	for secretKey, secretVal := range secretParams {
+		secretParam := v1alpha1.ParameterName(secretKey)
+		//Verify action must be reboot
+		if err := validateRebootAction(secretParam, secretVal, logger); err != nil {
+			return nil, err
+		}
+		if err := validateUniqueParam(fenceAgentParamNames, secretParam, logger); err != nil {
+			return nil, err
+		}
+		fenceAgentParams = appendParamToSlice(fenceAgentParams, secretParam, secretVal)
+	}
+
+	nodeName := getNodeName(far)
 	// append node parameters
-	nodeName := v1alpha1.NodeName(getNodeName(far))
 	for paramName, nodeMap := range far.Spec.NodeParameters {
-		if nodeVal, isFound := nodeMap[nodeName]; isFound {
-			fenceAgentParams = appendParamToSlice(fenceAgentParams, paramName, nodeVal)
+		if nodeVal, isFound := nodeMap[v1alpha1.NodeName(nodeName)]; isFound {
+			//Verify action must be reboot
+			if err := validateRebootAction(paramName, nodeVal, logger); err != nil {
+				return nil, err
+			}
+			// For node params we don't enforce uniqueness but use other value if defined, TODO explain why ?
```

Comment:
IIUC, we don't enforce uniqueness, and we allow creating a fence agent command with the same parameter. I think that the node params should override shared params (and we can add a log for having the node param priority)

Reply:
Furthermore, non-secret node params (i.e., `NodeParameters`) can't be duplicated with any secret param (i.e., `secretParams`).

Reply:
We can change the flow of building the command from SharedParams, SecretParams, and then NodeParams to NodeParams, SharedParams, and then SecretParams.
This way, the priority of the parameters should be:

1.  NodeParams - specific to a node and unique
2. SharedParams - shared in the cluster and not confidential.
3. SecretParams (from node) -  confidential for specific node and should not be explicit as NodeParams.
4. SecretParams (from default secret) - shared and used by default when not set by CR or node-specific secret

Reply:
>But secured params should not override public params of the same category, or the other way around. That would very likely be a configuration mistake, not?

Yep agreed.
In that case we'll get a validation error.

Reply:
Looks better with the new changes, and we can remove the `, TODO explain why ?` comment or just add the above priority of parameters in comments or (better IMO) in readme.

Reply:
In the current implementation:
a,f = allowed
b,c,d,e = not allowed

IIUC we agree on that implementation.
 
Current unit tests cover: a, b, c, d, e

Even though `f` is allowed I think there is a bug, where secure node params override all of the secure shared params (instead of just the duplicate ones).

 What's left: 
 - add unit test (which should confirm the bug I'm suspecting)
 - fix the bug
 
 Let me know if that makes sense.